a3p-integration/proposals/p:upgrade-19/test-lib/psm-helpers.js

```diff
@@ -0,0 +1,192 @@
+ * @property {import('@agoric/ertp').Amount<'nat'>} feePoolBalance
+ * @property {import('@agoric/ertp').Amount<'nat'>} mintedPoolBalance
+ * @property {import('@agoric/ertp').Amount<'nat'>} totalAnchorProvided
+ * @property {import('@agoric/ertp').Amount<'nat'>} totalMintedProvided
+ */
+
+const fromBoard = makeFromBoard();
+const marshaller = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+
+/**
+ * @param {string} path
+ */
+const objectFromVstorageEntries = async path => {
+  const rawEntries = await agoric.follow('-lF', `:${path}`, '-o', 'text');
+  return Object.fromEntries(marshaller.fromCapData(JSON.parse(rawEntries)));
+};
+
+export const snapshotAgoricNames = async () => {
+  const [brands, instances, vbankAssets] = await Promise.all([
+    objectFromVstorageEntries('published.agoricNames.brand'),
+    objectFromVstorageEntries('published.agoricNames.instance'),
+    objectFromVstorageEntries('published.agoricNames.vbankAsset'),
+  ]);
+  return { brands, instances, vbankAssets };
+};
+
+/**
+ * Similar to
+ * https://github.com/Agoric/agoric-3-proposals/blob/422b163fecfcf025d53431caebf6d476778b5db3/packages/synthetic-chain/src/lib/commonUpgradeHelpers.ts#L123-L139
+ * However, for an address that is not provisioned, `agoric wallet send` is
+ * needed because `agops perf satisfaction` hangs when trying to follow
+ * nonexistent vstorage path ":published.wallet.${address}".
+ *
+ * @param {string} address
+ * @param {Promise<string>} offerPromise
+ * @returns {Promise<SendOfferResult>}
+ */
+export const sendOfferAgoric = async (address, offerPromise) => {
+  const offerPath = await mkTemp('agops.XXX');
+  const offer = await offerPromise;
+  await fsp.writeFile(offerPath, offer);
+
+  const [settlement] = await Promise.allSettled([
+    execa({
+      all: true,
+    })`agoric wallet --keyring-backend=test send --offer ${offerPath} --from ${address} --verbose`,
+  ]);
+  return settlement.status === 'fulfilled'
+    ? settlement.value
+    : settlement.reason;
+};
+
+/**
+ * A variant of {@link sendOfferAgoric} that uses `agd` directly to e.g.
+ * control gas calculation.
+ *
+ * @param {string} address
+ * @param {Promise<string>} offerPromise
+ * @returns {Promise<SendOfferResult>}
+ */
+export const sendOfferAgd = async (address, offerPromise) => {
+  const offer = await offerPromise;
+  const networkConfig = await fetchEnvNetworkConfig({
+    env: process.env,
+    fetch,
+  });
+  const { chainName, rpcAddrs } = networkConfig;
+  const args = /** @type {string[]} */ (
+    // @ts-expect-error heterogeneous concat
+    [].concat(
+      [`--node=${rpcAddrs[0]}`, `--chain-id=${chainName}`],
+      [`--keyring-backend=test`, `--from=${address}`],
+      ['tx', 'swingset', 'wallet-action', '--allow-spend', offer],
+      '--yes',
+      '-bblock',
+      '-ojson',
+    )
+  );
+
+  const [settlement] = await Promise.allSettled([
+    execa('agd', args, { all: true }),
+  ]);
+
```

Comment:
I don't think this is different from
```suggestion
  const settlement = await execa('agd', args, { all: true }));
```

Reply:
After reviewing this change request, I realized that following the rebase to master, the file `psm-lib.js` had already been included in the upgrade-19 proposal's test library. To avoid duplicating code, I removed `psm-helpers.js` and updated psm-lib.js with the missing functions.

Regarding the implementation of `sendOfferAgd`, it was copied from [z:acceptance/test-lib/psm-lib.js](https://github.com/Agoric/agoric-sdk/blob/master/a3p-integration/proposals/z%3Aacceptance/test-lib/psm-lib.js#L470). The use of `Promise.allSettled()` was introduced in [this commit](https://github.com/Agoric/agoric-sdk/commit/236a61fb8a78fd97d8b1e7ac1c69e531c02de098) by Richard Gibson.

While I am unsure of the specific rationale behind this design, the change you suggested triggers an error because the expected structure of the settlement object is not provided.

For these reasons, I plan to leave psm-lib.js as is.

Reply:
It's different: `await` throws if the promise rejects. `allSettled` does not.

`sendOfferAgd` seems to report a rejection as a return value rather than throwing. Odd, but presumably expedient.